cvmfs/util/algorithm.cc

```diff
@@ -69,6 +76,129 @@ double StopWatch::GetTime() const {
 }
 
+namespace {
+
+static unsigned int CountDigits(uint64_t n) {
+  return (unsigned int)floor(log10(n) + 1);
+}
+
+static std::string GenerateStars(unsigned int n) {
+  return std::string(n, '*');
+}
+
+}  // anonymous namespace
+
+Log2Histogram::Log2Histogram(unsigned int nbins) {
+  assert(nbins != 0);
+  this->bins_.assign(nbins + 1, 0);  // +1 for overflow bin.
+  this->boundary_values_.assign(nbins + 1, 0);  // +1 to avoid big if statement
+
+  unsigned int i;
+  for (i = 1; i <= nbins; i++) {
+    this->boundary_values_[i] = (1 << ((i - 1) + 1));
+  }
+}
+
+std::vector<atomic_int32> UTLog2Histogram::GetBins(const Log2Histogram &h) {
+  return h.bins_;
+}
+
+std::string Log2Histogram::ToString() {
+  unsigned int i = 0;
+
+  unsigned int max_left_boundary_count = 1;
+  unsigned int max_right_boundary_count = 1;
+  unsigned int max_value_count = 1;
+  unsigned int max_stars = 0;
+  unsigned int max_bins = 0;
+  unsigned int total_stars = 38;
+  unsigned int total_sum_of_bins = 0;
+
+  for (i = 1; i <= this->bins_.size() - 1; i++) {
+    max_left_boundary_count = std::max(max_left_boundary_count,
+                                CountDigits(boundary_values_[i] / 2));
+    max_right_boundary_count = std::max(max_right_boundary_count,
+                                CountDigits(boundary_values_[i] - 1));
+    max_value_count = std::max(max_value_count, CountDigits(this->bins_[i]));
+    max_bins = std::max(max_bins, (unsigned int)
+                                atomic_read32(&(this->bins_[i])));
+    total_sum_of_bins += (unsigned int)atomic_read32(&(this->bins_[i]));
+  }
+
+  max_bins = std::max(max_bins, (unsigned int)atomic_read32(&(this->bins_[0])));
+  total_sum_of_bins += (unsigned int)atomic_read32(&(this->bins_[0]));
+
+  max_stars = max_bins * total_stars / total_sum_of_bins;
+
+  std::string format = " %" + StringifyUint(max_left_boundary_count < 2 ?
+                                  2 : max_left_boundary_count) +
+                  "d -> %" + StringifyUint(max_right_boundary_count) +
+                  "d :     %" + StringifyUint(max_value_count) + "d | %" +
+                  StringifyUint(max_stars) + "s |\n";
+
+  std::string title_format = " %" +
+                  StringifyUint((max_left_boundary_count < 2 ?
+                              2 : max_left_boundary_count) +
+                              max_right_boundary_count +
+                              4) +
+                  "s | %" + StringifyUint(max_value_count + 4) +
+                  "s | %" + StringifyUint(max_stars) + "s |\n";
+
+  std::string overflow_format = "%" +
+                  StringifyUint(max_left_boundary_count +
+                              max_right_boundary_count +
+                              5) +
+                  "s : %" + StringifyUint(max_value_count + 4) +
+                  "d | %" + StringifyUint(max_stars) + "s |\n";
+
+  std::string result_string = "";
+
+  const unsigned int kBufSize = 150;
```

Comment:
Actually: are we sure that the buffer size is big enough even for many (~20) bins?

Reply:
In principle, yes. The result histogram is stored in result_string. The buffer is needed for storing and appending each line in the histogram output. I can increase that to 200.